servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcUtils.java

```diff
@@ -263,10 +277,35 @@ static GrpcStatusException toGrpcException(Throwable cause) {
                 : new GrpcStatusException(toGrpcStatus(cause), () -> null);
     }
 
+    private static void validateStatusCode(HttpResponseStatus status) {
+        final int statusCode = status.code();
+        if (statusCode == OK.code()) {
+            return;
+        }
+        final GrpcStatusCode grpcStatusCode;
+        if (statusCode == BAD_GATEWAY.code() || statusCode == SERVICE_UNAVAILABLE.code() ||
+                statusCode == GATEWAY_TIMEOUT.code() || statusCode == TOO_MANY_REQUESTS.code()) {
+            grpcStatusCode = UNAVAILABLE;
+        } else if (statusCode == UNAUTHORIZED.code()) {
+            grpcStatusCode = UNAUTHENTICATED;
+        } else if (statusCode == FORBIDDEN.code()) {
+            grpcStatusCode = PERMISSION_DENIED;
+        } else if (statusCode == NOT_FOUND.code()) {
+            grpcStatusCode = UNIMPLEMENTED;
+        } else if (statusCode == BAD_REQUEST.code() || statusCode == REQUEST_HEADER_FIELDS_TOO_LARGE.code()) {
+            grpcStatusCode = INTERNAL;
```

Comment:
I'm surprised grpc-java maps these cases to `INTERNAL`.
In HTTP, bad request usually means that client sent something incorrectly rather than a server-side error occurred. I would expect `INVALID_ARGUMENT` in both of these cases. 

Also surprised that `412 Precondition Failed` is not mapped to `FAILED_PRECONDITION` and many other 4xx/5xx mappings may have a better grpc-status code.

Reply:
I took a stab at providing better mappings. I would suggest we can refine the http response code mapping in followup as well.